packages/client/src/components/pages/muster-page/table-unit-muster-compliance.tsx

```diff
@@ -0,0 +1,227 @@
+import {
+  Grid,
+  Paper,
+  Table,
+  TableBody,
+  TableContainer,
+  TableCell,
+  TableHead,
+  TableRow,
+} from '@material-ui/core';
+
+import * as Plotly from 'plotly.js';
+import React, {
+  useCallback,
+  useEffect,
+  useState,
+} from 'react';
+import Plot from 'react-plotly.js';
+
+import { UnitSelector } from '../../../selectors/unit.selector';
+import { Modal } from '../../../actions/modal.actions';
+import { formatErrorMessage } from '../../../utility/errors';
+import { UserSelector } from '../../../selectors/user.selector';
+import { MusterClient } from '../../../client/muster.client';
+import { useAppDispatch } from '../../../hooks/use-app-dispatch';
+import { useAppSelector } from '../../../hooks/use-app-selector';
+
+export const UnitMusterComplianceTable = (props: any) => {
+  const dispatch = useAppDispatch();
+  const units = useAppSelector(UnitSelector.all);
+  const trendChart = {
+    layout: {
+      height: 128,
+      autosize: true,
+      barmode: 'group',
+      yaxis: {
+        color: '#A9AEB1',
+        hoverformat: '.0f%',
+        showticklabels: true,
+        range: [0, 100],
+      },
+      xaxis: {
+        color: '#A9AEB1',
+        hoverformat: '',
+      },
+      hovermode: 'closest',
+      showlegend: false,
+      colorway: ['#7776AF', '#54BEA5', '#E87779', '#EACC73', '#81B2DA'],
+      margin: {
+        l: 40,
+        r: 20,
+        b: 20,
+        t: 20,
+      },
+    } as Partial<Plotly.Layout>,
+    config: {
+      displayModeBar: false,
+    } as Partial<Plotly.Config>,
+  };
+
+  const complianceBarChart = {
+    ...trendChart,
+  };
+
+  const [rosterFromDateIso, setRosterFromDateIso] = useState(props.rosterFromDateIso);
+  const [rosterToDateIso, setRosterToDateIso] = useState(props.rosterToDateIso);
+  const [rosterUnitId, setRosterUnitId] = useState(props.rosterUnitId);
+  const [unitComplianceData, setUnitComplianceData] = useState<Map<string, ComplianceData>>(new Map());
+  const [unitCompliancePlotData, setUnitCompliancePlotData] = useState<Map<string, Plotly.Data[]>>(new Map());
+
+  const org = useAppSelector(UserSelector.org)!;
+  const orgId = org.id;
+
+  //
+  // Effects
+  //
+
+  const showErrorDialog = useCallback((title: string, error: Error) => {
+    void dispatch(Modal.alert(`Error: ${title}`, formatErrorMessage(error)));
+  }, [dispatch]);
+
+  const reloadMusterComplianceData = useCallback(async () => {
+    try {
+      console.log(units);
+      console.log(rosterUnitId);
+      console.log(rosterFromDateIso);
+      // determine if we are displaying for all units or just a single unit
+      const unitsToDisplay = rosterUnitId === -1 ? units : [units.find(unit => unit.id === rosterUnitId)];
+      const plotData = new Map();
+      const complianceDataByUnit = new Map();
+      for (const unit of unitsToDisplay) {
```

Comment:
When I collapse this large block of code, then I see what it does. It "returns" these maps: `complianceDataByUnit` and `plotData`. I would prefer a function returning `complianceDataByUnit` and `plotData`. I would even move this function, a service, outside the the TSX module. 
<img width="1030" alt="Screen Shot 2021-08-11 at 11 24 27 AM" src="https://user-images.githubusercontent.com/58491416/129067292-33d0b707-727b-49e1-a612-adea38dbfdea.png">

Reply:
Good suggestion but I think I will leave as is for now. This logic is only used by this module and by this method. If that changes I will refactor then :)

Reply:
Good suggestion but I think I will leave as is for now. This logic is only used by this module and by this method. If that changes I will refactor then :)

Reply:
As a learning conversation...Would it be a good idea for an API/service to do most of this heavy lifting instead of doing it in the UI client? The API would return appropriate and directly consumable data structures.